include/mqtt/message.hpp

```diff
@@ -529,46 +528,48 @@ class basic_publish_message {
         }
     }
 
-    template <typename Iterator>
-    basic_publish_message(Iterator b, Iterator e) {
-        if (b + 1 > e) throw remaining_length_error();
-        fixed_header_ = static_cast<std::uint8_t>(*b);
+    basic_publish_message(buffer buf) {
+        if (buf.empty())  throw remaining_length_error();
+        fixed_header_ = static_cast<std::uint8_t>(buf.front());
         auto qos = publish::get_qos(fixed_header_);
-        ++b;
+        buf = buf.substr(1);
```

Comment:
You're not having the basic_publish_message object actually take ownership of the data that it's referring to.

So unless you plan to have basic_publish_message hold onto an instance of the passed in mqtt::buffer, this function should instead take an mqtt::string_view (by value), to avoid all of the lifetime management activity that's going on here.

Reply:
See https://github.com/redboltz/mqtt_cpp/pull/337#discussion_r314573965